src/java/org/apache/cassandra/config/Properties.java

```diff
@@ -0,0 +1,215 @@
+        return sb;
+    }
+
+    public static CharSequence toStringTree(Map<String, Property> input)
+    {
+        SortedMap<String, Property> properties = new TreeMap<>(input);
+        StringBuilder sb = new StringBuilder();
+        addConfigs(defaultLoader(), properties, sb, 0);
+        return sb;
+    }
+
+    private static void addConfigs(Loader loader, SortedMap<String, Property> properties, StringBuilder sb, int layer)
+    {
+        for (Map.Entry<String, Property> e : properties.entrySet())
+        {
+            for (int i = 0; i < layer; i++)
+                sb.append("  ");
+            sb.append(e.getKey()).append(":");
+            Property property = e.getValue();
+            Map<String, Property> subProperties = Properties.isPrimitive(property) || Properties.isCollection(property) ? Collections.emptyMap() : loader.getProperties(property.getType());
+            if (subProperties.isEmpty())
+            {
+                sb.append(" ").append(normalize(property.getType()));
+                sb.append("\n");
+            }
+            else
+            {
+                sb.append(" # ").append(normalize(property.getType())).append("\n");
+                addConfigs(loader, new TreeMap<>(subProperties), sb, layer + 1);
+            }
+        }
+    }
+
+    private static String normalize(Class<?> type)
+    {
+        if (type.isEnum())
+            return "Enum<" + type.getSimpleName() + ">";
+        if (Number.class.isAssignableFrom(type))
+        {
+            if (Integer.class.equals(type))
+                return "int";
+            if (Long.class.equals(type))
+                return "long";
+            if (Float.class.equals(type))
+                return "float";
+            if (Double.class.equals(type))
+                return "double";
+        }
+        return type.getSimpleName();
+    }
+
+    private static final class AndThen extends Property
+    {
+        private final Property root;
+        private final Property leaf;
+
+        AndThen(Property root, Property leaf, String delimiter)
+        {
+            super(root.getName() + delimiter + leaf.getName(), leaf.getType());
+            this.root = root;
+            this.leaf = leaf;
+        }
+
+        @Override
+        public Class<?>[] getActualTypeArguments()
+        {
+            return leaf.getActualTypeArguments();
+        }
+
+        @Override
+        public void set(Object o, Object o1) throws Exception
+        {
+            Object parent = root.get(o);
+            if (parent == null)
+                return;
+            leaf.set(parent, o1);
+        }
+
+        @Override
+        public Object get(Object o)
+        {
+            try
+            {
+                Object parent = root.get(o);
+                if (parent == null)
+                    return null;
+                return leaf.get(parent);
+            }
+            catch (Exception e)
+            {
+                if (!(root instanceof AndThen))
```

Comment:
if you have a chain like `foo.bar.baz` only log this on the top level (`root` isn't an `AndThen`)